### PR TITLE
WIP: #121

### LINKED ISSUE #121: Generalize all retry methods under a uniform interface
- [x] Design generalization
 - [x] Implement sketch of generalization
 - [x] Add a couple tests to flesh out design
 - [x] Replace specialized retry methods with generic combinators, to further flesh out design
 - [ ] Update documentation, including README.md, adding a section on composing retry strategies

Proposed design now sketched out in #138

### PR DESCRIPTION
**Pull request by jdegoes**
Fixes #121, eventually. Maybe.

### CODE REVIEW

**Comment by jdegoes** [comment]
@ktonga 

Please take a look at this WIP for #121. You can submit PRs to my fork if you like and I'll merge them in.

I'll add some of my own thoughts to this PR.

File: core/shared/src/main/scala/scalaz/zio/Retry.scala [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+   * A retry strategy that always retries and counts the number of retries.
+   */
+  final def counted[E]: Retry[Int, E] =
+    Retry[Int, E](IO.now(0), (_, i) => IO.now(i + 1))
+
+  /**
+   * A retry strategy that always retries and computes the time since the
+   * beginning of the process.
+   */
+  final def timed[E]: Retry[(Long, Long), E] = {
```
Comment (jdegoes) [inline core/shared/src/main/scala/scalaz/zio/Retry.scala]: This is the one thing I don't like. All the state has to be exposed via `S`, which means you get this nasty tuple here that carries a piece of state (the starting time) that I don't want to expose. Maybe it'd be interesting to look at having a part of the state hidden, and then a projection from the hidden state `S0 => S` (to the visible part, which would be exposed in functions like `whileState`).

**Comment by ktonga** [comment]
@jdegoes are you gonna keep working on it? you wanna kinda split the work somehow so we don't step on each other toes?

File: core/shared/src/main/scala/scalaz/zio/IO.scala [hunk 2]
```diff
@@ -368,11 +368,28 @@ sealed abstract class IO[+E, +A] { self =>
   final def zipWith[E1 >: E, B, C](that: IO[E1, B])(f: (A, B) => C): IO[E1, C] =
     self.flatMap(a => that.map(b => f(a, b)))
 
+  /**
+   * Sequentially zips this effect with the specified effect, combining the
+   * results into a tuple.
+   */
+  final def zip[E1 >: E, B](that: IO[E1, B]): IO[E1, (A, B)] =
```
Comment (sideeffffect) [inline core/shared/src/main/scala/scalaz/zio/IO.scala]: shouldn't we call it `seq` (as a dual to `par`)?

Comment (jdegoes) [reply core/shared/src/main/scala/scalaz/zio/IO.scala]: Good idea, in which case we should rename `zipWith` => `seqWith`, and maybe add `parWith` for symmetry.

**Comment by jdegoes** [comment]
@ktonga I'm happy to hand this off to you, if you have the bandwidth to finish it.

**Comment by ktonga** [comment]
@jdegoes Yes, I do need it for my work so I can spend a full day or to on it.

**Comment by ktonga** [comment]
You wanna close this PR? I can branch off of your master and open a new PR to avoid cascade merging

**Comment by jdegoes** [comment]
@ktonga Awesome, will close!

**Pull request closed without merging**